cmd/openfga/run.go

```diff
@@ -0,0 +1,773 @@
+
+		go func() {
+			var err error
+			if config.HTTP.TLS.Enabled {
+				if config.HTTP.TLS.CertPath == "" || config.HTTP.TLS.KeyPath == "" {
+					logger.Fatal("'http.tls.cert' and 'http.tls.key' configs must be set")
+				}
+				err = httpServer.ListenAndServeTLS(config.HTTP.TLS.CertPath, config.HTTP.TLS.KeyPath)
+			} else {
+				err = httpServer.ListenAndServe()
+			}
+			if err != http.ErrServerClosed {
+				logger.Fatal("HTTP server closed with unexpected error", zap.Error(err))
+			}
+		}()
+		logger.Info(fmt.Sprintf("HTTP server listening on '%s'...", httpServer.Addr))
+	}
+
+	var playground *http.Server
+	if config.Playground.Enabled {
+		if !config.HTTP.Enabled {
+			return errors.New("the HTTP server must be enabled to run the openfga playground")
+		}
+
+		authMethod := config.Authn.Method
+		if !(authMethod == "none" || authMethod == "preshared") {
+			return errors.New("the playground only supports authn methods 'none' and 'preshared'")
+		}
+
+		playgroundAddr := fmt.Sprintf(":%d", config.Playground.Port)
+		logger.Info(fmt.Sprintf("🛝 starting openfga playground on http://localhost%s/playground", playgroundAddr))
+
+		tmpl, err := template.ParseFS(assets.EmbedPlayground, "playground/index.html")
+		if err != nil {
+			return fmt.Errorf("failed to parse playground index.html as Go template: %w", err)
+		}
+
+		fileServer := http.FileServer(http.FS(assets.EmbedPlayground))
+
+		policy := backoff.NewExponentialBackOff()
+		policy.MaxElapsedTime = 3 * time.Second
+
+		var conn net.Conn
+		err = backoff.Retry(
+			func() error {
+				conn, err = net.Dial("tcp", config.HTTP.Addr)
+				return err
+			},
+			policy,
+		)
+		if err != nil {
+			return fmt.Errorf("failed to establish playground connection to HTTP server: %w", err)
+		}
+
+		playgroundAPIToken := ""
+		if authMethod == "preshared" {
+			playgroundAPIToken = config.Authn.AuthnPresharedKeyConfig.Keys[0]
+		}
+
+		mux := http.NewServeMux()
+		mux.Handle("/", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+
+			if strings.HasPrefix(r.URL.Path, "/playground") {
+				if r.URL.Path == "/playground" || r.URL.Path == "/playground/index.html" {
+					err = tmpl.Execute(w, struct {
+						HTTPServerURL      string
+						PlaygroundAPIToken string
+					}{
+						HTTPServerURL:      conn.RemoteAddr().String(),
+						PlaygroundAPIToken: playgroundAPIToken,
+					})
+					if err != nil {
+						w.WriteHeader(http.StatusInternalServerError)
+						logger.Error("failed to execute/render the playground web template", zap.Error(err))
+					}
+
+					return
+				}
+
+				fileServer.ServeHTTP(w, r)
+				return
+			}
+
+			http.NotFound(w, r)
+		}))
+
+		playground = &http.Server{Addr: playgroundAddr, Handler: mux}
+
+		go func() {
+			err = playground.ListenAndServe()
+			if err != http.ErrServerClosed {
+				logger.Fatal("failed to start the openfga playground server", zap.Error(err))
+			}
+			logger.Info("shutdown the openfga playground server")
+		}()
+	}
+
+	logger.Info(
```

Comment:
Can we move this up before we actually start the OpenFGA server?

Reply:
Yep. Done. Thanks!